CORDA-3600 Add flowIO request to checkpoint

williamvigorr3 opened the pull request:
FlowMonitor updates the checkpoint database with stuck flow information instead of simply logging.

williamvigorr3 committed:
Update Checkpoint DB to update flow io request

williamvigorr3 committed:
Modify flow monitor to update Checkpoint DB with waiting flows

This happens periodically.

williamvigorr3 committed:
Refactored code to avoid looping twice and updated tests

williamvigorr3 committed:
Fix tests after rebasing

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
+                ioRequest?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+                if (ioRequest != null) {
+                    checkpointStorage.updateFlowIoRequest(flow.id, ioRequest)
+                }
```
This can go into the `let` block above:

```kotlin
flow.ioRequest().ioRequest?.let {
    logger.info(warningMessageForFlowWaitingOnIo(it, flow, suspensionDuration)) }
    checkpointStorage.updateFlowIoRequest(flow.id, it)
}
```

williamvigorr3 replied:
Done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt:
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
     }
 
+    @Test(timeout = 300_000)
+    fun `Check that flow io request can be added to a checkpoint`() {
```
Rename the test to:

`checkpoint can be updated with flow io request information`

williamvigorr3 replied:
Done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt:
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
+            val serializedFlowState = checkpoint.flowState.checkpointSerialize(context = CheckpointSerializationDefaults.CHECKPOINT_CONTEXT)
+            checkpointStorage.addCheckpoint(id, checkpoint, serializedFlowState)
+            val checkpoint = checkpointStorage.getDBCheckpoint(id)
+            assertNull(checkpoint?.ioRequestType)
```
Do !! to make sure it is not the checkpoint that is null instead of the `ioRequestType`

williamvigorr3 replied:
Oops done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/persistence/DBCheckpointStorageTests.kt:
```diff
@@ -347,6 +350,27 @@ class DBCheckpointStorageTests {
+            val checkpoint = checkpointStorage.getDBCheckpoint(id)
+            assertNotNull(checkpoint?.ioRequestType)
+            val classFromCheckpoint = checkpoint?.ioRequestType!!
+            assert(classFromCheckpoint == FlowIORequest.Sleep::class.java)
```
Can we use `assertEquals(FlowIORequest.Sleep::class.java, classFromCheckpoint)`

williamvigorr3 replied:
Done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -69,6 +73,17 @@ class FlowFrameworkTests {
     private lateinit var notaryIdentity: Party
     private val receivedSessionMessages = ArrayList<SessionTransfer>()
 
+    private fun dbCheckpointStorageWithStubPerfRecorder() : DBCheckpointStorage {
```
Just call it `checkpointStorage` no one will care about the performance recorder.

lankydan replied:
And store it in a variable

williamvigorr3 replied:
Done that.

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
             val terminationSignal = Semaphore(0)
```
Can we give the test a better name.

williamvigorr3 replied:
Done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
+            bobFlowMonitor.updateDbAndLogWaitingFlows()
+
+            aliceNode.database.transaction {
+                assert(dbCheckpointStorageWithStubPerfRecorder().getCheckpointIoRequest(aliceFlowId) == FlowIORequest.Receive::class.java)
```
Use `assertEquals`

williamvigorr3 replied:
Done.

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -166,46 +181,83 @@ class FlowFrameworkTests {
     }
 
     @Test(timeout=300_000)
 	fun `FlowMonitor flow is running`() { // flow is running a "take a long time" task
```
Can we give this a better name as well please.

williamvigorr3 replied:
Done this..

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -227,6 +228,17 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
+    }
+
+    @VisibleForTesting
+    fun getDBCheckpoint(id: StateMachineRunId): DBFlowCheckpoint? {
```
I personally would rather we just didn't have this function, but I am aware that makes some duplicated code in testing classes but then it isnt too bad.

mnesbit replied:
What tests actually need this?

williamvigorr3 replied:
Some of the `FlowFrameworkTests` that use the FlowMonitor and a test I added to`DBCheckpointStorageTests`.  Actually this would be easy to remove (this used to be needed before @lankydan made some changes to #600 which I overlooked when rebasing).

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
-            flow.ioRequest()?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+    @VisibleForTesting
+    fun updateDbAndLogWaitingFlows() {
+        database.transaction {
```
I think it makes sense to use a single database transaction (like you have done) 👍

lankydan reviewed (commented):
Looks about right. Incoming tidy up comments!

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowMonitor.kt:
```diff
@@ -53,9 +57,16 @@ internal class FlowMonitor(
+            for ((flow, suspensionDuration) in waitingFlowDurations(suspensionLoggingThreshold)) {
+                val ioRequest = flow.ioRequest()
+                ioRequest?.let { request -> logger.info(warningMessageForFlowWaitingOnIo(request, flow, suspensionDuration)) }
+                if (ioRequest != null) {
```
We should probably also retrieve checkpoint to test if the `ioRequest` has changed (is `null` or something else`). That way we can remove an extra update. It should be retrieving it from the hibernate cache so performance _should_ be alright. That being said, I am not sure if this will improve the performance or not... (I don't know hibernate performance well enough).

williamvigorr3 replied:
Okay I have done this in DBCheckpointStorage (as we were already extracting and updating there).

williamvigorr3 committed:
Fix MR comments (non-functional refactor of tests + FlowMonitor).

williamvigorr3 committed:
Made visible for testing method private in DBCheckpointStorage

This is not needed anymore.

williamvigorr3 committed:
Explicity check if ioRequestType has changed in update method

williamvigorr3 committed:
Fix shadowing warning

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -250,6 +257,10 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
         }
     }
 
+    private fun getDBCheckpoint(id: StateMachineRunId): DBFlowCheckpoint? {
```
I think this should use `find` instead of `get`. You can then use it in some of the other functions.

williamvigorr3 replied:
This should be done now.

lankydan commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowFrameworkTests.kt:
```diff
@@ -161,51 +176,87 @@ class FlowFrameworkTests {
+                assert(dbCheckpointStorage.getCheckpointIoRequest(aliceFlowId) == FlowIORequest.Receive::class.java)
+            }
+            bobNode.database.transaction {
+                assert(dbCheckpointStorage.getCheckpointIoRequest(bobFlowId) == FlowIORequest.Receive::class.java)
```
`assertEquals`

williamvigorr3 replied:
Oops sorry I thought I fixed all of those. Done now.

kyriathar commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -228,8 +228,10 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
         val checkpoint = getDBCheckpoint(id)
-        checkpoint?.ioRequestType = ioRequest.javaClass
-        currentDBSession().update(checkpoint)
+        if (checkpoint?.ioRequestType != ioRequest) {
```
Really minor thing but would it maybe be better to write this `if` block as `checkpoint?.let` { ...

williamvigorr3 replied:
Yeah I have done this. I think I understood what you meant but if there is a cleaner way let me know.

kyriathar replied:
Yes you are right in that its not very clear, if we could please replace `currentDBSession().update(checkpoint)` with `currentDBSession().update(it)  `. 
To make it a bit more clear, maybe, it could be written like: `it.apply {ioRequestType = ioRequest.javaClass}` i think.

williamvigorr3 committed:
Import non deprecated Assert into test

williamvigorr3 committed:
Use AssertEquals not assert in test

williamvigorr3 committed:
Address more comments (minor refactor) of DBCheckpointStorage

williamvigorr3 committed:
Minor fix use it instead of referencing object explicitly

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -227,6 +226,16 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
 
+    override fun updateFlowIoRequest(id: StateMachineRunId, ioRequest: FlowIORequest<*>) {
+        val checkpoint = getDBCheckpoint(id)
+        checkpoint?.let {
```
Makes sense that this also does:

```kotlin
val entity = requireNotNull(getDBCheckpoint(id)) {
            "Checkpoint with id: ${id.uuid} does not exist in database."
        }
```

Like you did in `updateDBCheckpoint`. The current null check can then be removed.

williamvigorr3 replied:
Done.

williamvigorr3 committed:
Add null check to DBCheckpointStorage

mnesbit commented:
I am a bit confused by this. FlowIO requests are captured in the stack of the fiber. I can understand needing to put some tracking strings in the DB, but do we really want to serialize/deserialize objects to the DB just for informational purposes? Definitely anything we write to DB is trumped by the info in the fiber.

lankydan commented:
We are writing the flow io request to the database to provide something that users can query for. I agree that the information in the fiber is the _correct_ information but that requires any filtering based on the flow io request to be done in memory.

Not doing this work in-memory is why we have done it in the `FlowMonitor`. Only filling in the flow io request field of flows that have been suspended for longer than a period of time.

We shouldn't be doing a serialization/deserialization here. We are determining whether we need to update the flow io request using in-memory data and then updating only that column in the table.

If we want to switch to doing this all in memory (which will make it 100% accurate) we need to decide sooner rather than later due to the effects querying the data.

williamvigorr3 committed:
Revert changes to Flow Monitor.

We will instead store the information in the main thread of the
state machine.

williamvigorr3 committed:
Remove now uneeded API and make statemachine update ioRequest

williamvigorr3 committed:
Add Integration Test to check statemachine updates DB on Recieve

rick-r3 commented:
I thought we were going to flip the flow IO column to be a string and have the class short name ?

lankydan commented:
@rick-r3 I talked to @roastario about this, and from his perspective, it does not matter whether it is a class name or a string. We have stuck with the `Class` since it makes the code tidier. We can always switch over to a string later on if stefano's implementation is hindered by doing so.

mnesbit commented:
I am worried that Class might not be reloadable after upgrades, so we might lose the info for the incompatible flows.

lankydan commented:
Sounds like we should switch to a string then ^^

williamvigorr3 committed:
Use simpleName in checkpoint storage instead of class.

Hibernate was previously resetting the class field this is now
set to null (when getting checkpoint form DB) and a new method
for getting back the simple name as a string.

williamvigorr3 commented:
Okay I have switched to a String using simpleName.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -336,7 +345,7 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
             this.compatible = checkpoint.compatible
             this.progressStep = checkpoint.progressStep
-            this.ioRequestType = checkpoint.flowIoRequest
+            this.ioRequestType = checkpoint.flowIoRequest?.simpleName?.take(MAX_SIMPLE_NAME_LENGTH)
```
I think because we are in control of the names of the `FlowIORequest` classes that we don't need to worry about the length being hit. All of the current instances of `FlowIORequest` should be under the limit. If we manage to have a super long length name in the future we can change this code then (but i doubt we would manage to do that).

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -382,7 +391,7 @@ class DBCheckpointStorage(private val checkpointPerformanceRecorder: CheckpointP
             status = status,
             progressStep = progressStep,
-            flowIoRequest = ioRequestType,
+            flowIoRequest = null,
```
Lets just make `flowIoRequest` type a `String` and set it as such so we don't have problems loading from the database.

lankydan commented on node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt:
```diff
@@ -41,4 +42,10 @@ interface CheckpointStorage {
+     * Get the simpleName of the flowIoRequest which the flow suspended on. Returns null if the flow didn't suspend
+     * on an flowIoRequest.
+     */
+    fun getFlowIoRequest(id: StateMachineRunId): String?
```
We can remove this once `Checkpoint.flowIORequest` has become a string

williamvigorr3 committed:
Update StateMachineState to store simple name.

williamvigorr3 committed:
Fix after rebase broke stuff + renamed test

williamvigorr3 committed:
Fix Detekt issue

williamvigorr3 committed:
Remove uneeded null assertion

lankydan reviewed (approved):
I think we have made it!

williamvigorr3 commented:
:champagne:
